python/ray/train/_internal/worker_group.py

```diff
@@ -360,26 +361,34 @@ def add_workers(self, num_workers: int):
         for i in range(len(new_actors)):
             self.workers.append(Worker(actor=new_actors[i], metadata=metadata[i]))
 
-    def _move_workers_with_ip_to_front(self, ip):
-        # Hack to avoid OOMs.
-        # This is just a temporary solution for Train loading entire checkpoints
-        # into memory by ensuring that the rank 0 worker is on the same node as
-        # trainable, thus allowing for lazy checkpoint transfer to be used.
-        # See https://github.com/ray-project/ray/issues/33073
-        # for more context.
-        # TODO remove
-        workers_with_ip = []
-        indices_to_remove = set()
-        for i, worker in enumerate(self.workers):
-            if worker.metadata.node_ip == ip:
-                workers_with_ip.append(worker)
-                indices_to_remove.add(i)
-        if workers_with_ip:
-            self.workers = workers_with_ip + [
-                worker
-                for i, worker in enumerate(self.workers)
-                if i not in indices_to_remove
-            ]
+    def group_workers_by_ip(self, _first_ip: Optional[str] = None):
+        """Groups workers by IP.
+
+        This is useful for collocating workers on the same node.
+
+        Args:
+            _first_ip: The first IP to group by.
+                Hack to avoid OOMs.
+                This is just a temporary solution for Train loading entire checkpoints
+                into memory by ensuring that the rank 0 worker is on the same node as
+                trainable, thus allowing for lazy checkpoint transfer to be used.
+                See https://github.com/ray-project/ray/issues/33073
+                for more context.
+                TODO remove this argument.
+        """
+        ip_to_workers = defaultdict(list)
```

Comment:
Should we use `OrderedDict` here to maintain the insertion order? so that the `_first_ip` can always be the first?

Reply:
Oh technically dictionaries follow insertion order as of Python 3.7.